### PR TITLE
SLING-9695 Added options for datastore

### PR DESCRIPTION
**Pull request by cmrockwell**
Added options for datastore sling.run.mode.install.options, which could be combined with either oak_mongo or oak_tar. Updated Readme.md describing install options.

### CODE REVIEW

**Review by rombert — CHANGES_REQUESTED** [review]
Thanks for the PR @cmrockwell . There are conflicts here, due to the fact that we switched to the feature model.

I would suggest that you add an additional feature file, maybe under src/main/features/oak/datastore which configures the FileDataStore.

Users could then launch it using something like

    java -jar target/dependency/org.apache.sling.feature.launcher.jar -f target/slingfeature-tmp/feature-oak_mongo.json,target/slingfeature-tmp/feature-oak_datastore_file.json

The only think that I'm not sure how to do is to configure the SegmentNodeStore to use a custom blob store. Maybe defined a variable "oak.customBlobStore" that is "false" by default and that can be overridden from the CLI using -V oak.customBlobStore=true ?

**Comment by cmrockwell** [comment]
@rombert Thanks for the feedback. Launching the instance as you suggest with a series of feature.json is a bit cumbersome as compared to the prior way (`java -jar org.apache.sling.starter-12-SNAPSHOT.jar`). I worry about typing it wrong or missing one or more important features; this introducing bugs. Is there is a maven command (and pom config) to generate Feature Archive as a Jar file containing a predefined set of features (oak_tar, file datastore, composum, web console) that could be started without all the -f args?

**Comment by cmrockwell** [comment]
Update: I updated the pull request resolving the PM/FM conflict and created an aggregate feature file (oak_tar_fds.json) This satisfies my requirement for keeping the launch command relatively simple. Also tweaked the README to make it even more useful.

**Comment by cmrockwell** [comment]
One note. The path was hard coded as shown below
`        "org.apache.jackrabbit.oak.plugins.blob.datastore.FileDataStore": {
            "minRecordLength":4096,
            "path":"sling/repository/datastore",
            "cacheSizeInMB":128
        }`
boot.json uses a variable ${sling.home} but when I tried using that for the datastore `path` it created a directory with that variable name (not the value). It created a directory literally called "${sling.home}"

**Comment by rombert** [comment]
Thanks for the update. I am not sure we should be generating so many aggregates ... we have a large number of combinations of NodeStores and DataStores, e.g.

- SegmentNodeStore
- DocumentNodeStore + MongoDB
- DocumentNodeStore + RDB (Postgres?)

- No BlobStore
- FileBlobStore
- MongoBlobStore
- RDBBlobStore
(etc, etc)

The idea behind having an additional file which configures just the DataStore was to combine one NodeStore with one DataStore, instead of 'hardcoding' combinations. Additionally, if we generate `far` files as well, we greatly increase the storage needed. But I should think about this some more tomorrow, it's a bit late over here right now :-)

About the hardcoded path, I think you can work around it by defining another variable

```
    "variables": {
        "repository.home": "${sling.home}/repository"
    }
```

And then referencing `${repository.home}` in the configuration values.

**Comment by rombert** [comment]
Oh, an according to https://github.com/apache/sling-org-apache-sling-feature-launcher/blob/7aa00ee307f5b3452b6a1cc3f3e2959251a3d6ca/src/main/java/org/apache/sling/feature/launcher/impl/Bootstrap.java#L73 , a variable named `sling.home` should be supported, please file a bug.

**Comment by cmrockwell** [comment]
I don't necessarily see a problem adding any or all of aggregates @rombert IMO it's nice to have working options and examples. Nevertheless ticket SLING-9695 requests file datastore as default, so I went ahead and removed that oak_tar_fds aggregate and configured file datastore by default. Many application developers would want the performance boost that comes with saving binaries outside of the segment (or document) store, so one of the example aggregates should demonstrate that.

File: README.md [hunk 1]
```diff
@@ -20,17 +20,26 @@ See [Releasing a new version of the Sling starter](https://cwiki.apache.org/conf
         directory. It is advisable to use a work directory outside of the
         project directory.
 
-1) Build the Sling Starter using 
+1) Build the Sling Starter using `mvn clean install` in the current directory.
+2) Copy the launch to project directory `cp target/dependency/org.apache.sling.feature.launcher.jar $path-to-project`
```
Comment (rombert) [inline README.md:24]: Any particular reason to add this step? I think it's simpler to not copy it, but use it directly in its configured location.

File: README.md [hunk 2]
```diff
@@ -20,17 +20,26 @@ See [Releasing a new version of the Sling starter](https://cwiki.apache.org/conf
...
-in the current directory.
 
-2) Start Sling backed by an Oak SegmentStore with
+Example 1: Start an instance using the launcher specifying a feature archive (.far)
+This example uses the aggregate `oak_tar_far` as defined in the pom.xml
 
-    java -jar target/dependency/org.apache.sling.feature.launcher.jar -f target/slingfeature-tmp/feature-oak_tar.json
-	
-3) Browse Sling in:
+        java -jar org.apache.sling.feature.launcher.jar -f org.apache.sling.starter-12-SNAPSHOT-oak_tar_far.far
```
Comment (rombert) [inline README.md:31]: You are missing the `target/slingfeature-tmp` prefix here.

File: README.md [hunk 3]
```diff
@@ -20,17 +20,26 @@ See [Releasing a new version of the Sling starter](https://cwiki.apache.org/conf
...
 
-    java -jar target/dependency/org.apache.sling.feature.launcher.jar -f target/slingfeature-tmp/feature-oak_tar.json
-	
-3) Browse Sling in:
+        java -jar org.apache.sling.feature.launcher.jar -f org.apache.sling.starter-12-SNAPSHOT-oak_tar_far.far 
+	 
+Example 2: Start an instance using the launcher specifying an aggregate feature file. 
+This example uses the aggregate `feature-oak_tar.json` as defined in the pom.xml
+
+        java -jar org.apache.sling.feature.launcher.jar -f feature-oak_tar.json
```
Comment (rombert) [inline README.md:36]: You are missing the `target/slingfeature-tmp` prefix here.

File: README.md [hunk 4]
```diff
@@ -20,17 +20,26 @@ See [Releasing a new version of the Sling starter](https://cwiki.apache.org/conf
...
+
+Example 3: Start an instance using the launcher specifying a set feature files.
+ 
+        java -jar org.apache.sling.feature.launcher.jar -f feature-base.json,feature-boot.json,....
```
Comment (rombert) [inline README.md:40]: Perhaps it would be better to point to the feature model launcher documentation instead?

File: pom.xml [hunk 5]
```diff
@@ -127,7 +127,7 @@
                             <classifier>oak_mongo_far</classifier>
                             <includeClassifier>oak_mongo</includeClassifier>
                         </archive>
-                    </archives>
```
Comment (rombert) [inline pom.xml:130]: Can you please drop these whitespace-only changes?

**Review by rombert — CHANGES_REQUESTED** [review]
Agree, it makes sense to switch to datastore by default. Added some minor notes, it would be great if you could address them.

**Comment by cmrockwell** [comment]
This pull request has expired.

**Pull request closed without merging**